From 19a1ba3305a560ab03dacc010177b02cf06ecca6 Mon Sep 17 00:00:00 2001
From: Jelle van der Waa <jvanderwaa@redhat.com>
Date: Tue, 8 Sep 2026 09:45:19 +0200
Subject: [PATCH] Apply `--no-dereference` to chmod when supported

This option exists since coreutils 9.5 which RHEL 9.9 and RHEL 8.10
don't have so gracefully fall back to the old behaviour.

Fixes: COCKPIT-1900
---
 src/dialogs/permissions.tsx | 27 +++++++++++++++++-----
 test/check-application      | 45 +++++++++++++++++++++++++++++++++++++
 test/reference              |  2 +-
 3 files changed, 67 insertions(+), 7 deletions(-)

diff --git a/src/dialogs/permissions.tsx b/src/dialogs/permissions.tsx
index 3959fe06d..5990dc7a7 100644
--- a/src/dialogs/permissions.tsx
+++ b/src/dialogs/permissions.tsx
@@ -19,7 +19,7 @@ import { TextInput } from '@patternfly/react-core/dist/esm/components/TextInput'
 import { Stack } from '@patternfly/react-core/dist/esm/layouts/Stack';
 
 import cockpit from 'cockpit';
-import type { BasicError } from 'cockpit';
+import { type BasicError, type SpawnOptions } from 'cockpit';
 import { InlineNotification } from 'cockpit-components-inline-notification';
 import type { Dialogs, DialogResult } from 'dialogs';
 import { useInit } from 'hooks';
@@ -104,6 +104,24 @@ function mode_to_args(mode: number) {
     return chmod_args.join(",");
 }
 
+async function chmod(paths: string[], mode: string, recursive: boolean,
+    spawn_options: SpawnOptions & { binary?: false }) {
+    const options = [];
+    if (recursive) {
+        options.push("--recursive");
+    }
+
+    // Only CoreUtils 9.5 supports --no-dereference
+    try {
+        await cockpit.spawn(["chmod", "--no-dereference", "--version"], { ...spawn_options, environ: ["LC_ALL=C"] });
+        options.push("--no-dereference");
+    } catch (err) {
+        console.warn("No support for '--no-dereference'", err);
+    }
+
+    await cockpit.spawn(["chmod", ...options, "--", mode, ...paths], spawn_options);
+}
+
 const EditPermissionsModal = ({ dialogResult, items, path } : {
     dialogResult: DialogResult<void>,
     items: FolderFileInfo[],
@@ -167,9 +185,7 @@ const EditPermissionsModal = ({ dialogResult, items, path } : {
 
     const spawnEncloseFiles = async () => {
         try {
-            await cockpit.spawn(["chmod", "--recursive", "--", mode_to_args(mode), full_path],
-                                { superuser: "try", err: "message" });
-
+            await chmod([full_path], mode_to_args(mode), true, { superuser: "try", err: "message" });
             await cockpit.spawn(["chown", "--recursive", "--no-dereference", "--", owner + ":" + group, full_path],
                                 { superuser: "try", err: "message" });
 
@@ -188,8 +204,7 @@ const EditPermissionsModal = ({ dialogResult, items, path } : {
 
         try {
             if (permissionChanged)
-                await cockpit.spawn(["chmod", "--", mode.toString(8), ...file_paths],
-                                    { superuser: "try", err: "message" });
+                await chmod(file_paths, mode.toString(8), false, { superuser: "try", err: "message" });
 
             if (ownerChanged)
                 await cockpit.spawn(["chown", "--no-dereference", "--", owner + ":" + group, ...file_paths],
diff --git a/test/check-application b/test/check-application
index c7ddf1db3..52fa1b727 100755
--- a/test/check-application
+++ b/test/check-application
@@ -1706,6 +1706,51 @@ class TestFiles(testlib.MachineCase):
 
         self.chdir('/home/admin')
 
+        # Don't dereference symlinks in directories, neither chmod/chown allow this
+        m.execute("""
+        touch /tmp/passwd
+        chmod 600 /tmp/passwd
+        runuser -u admin -- mkdir -p /home/admin/recursive/a
+        runuser -u admin -- ln -s /tmp/passwd /home/admin/recursive/b
+        """)
+        self.addCleanup(m.execute, "rm -f /tmp/passwd")
+
+        open_permissions_dialog("recursive")
+        b.wait_in_text("#edit-permissions-group", "admin")
+        select_access("read-write")
+        b.select_from_dropdown("#edit-permissions-group", "testuser")
+        b.click(".pf-v6-c-modal-box button.pf-m-secondary")
+        b.wait_not_present(".pf-v6-c-modal-box")
+
+        b.wait_in_text("[data-item='recursive'] .item-owner", "admin:testuser")
+        self.assertEqual(self.stat("%G", "/tmp/passwd"), "root")
+        self.assertEqual(self.stat("%A", "/tmp/passwd"), "-rw-------")
+
+        m.execute("""
+        touch /home/admin/passwd
+        chown admin: /home/admin/passwd
+        """)
+        b.wait_in_text("[data-item='passwd'] .item-owner", "admin")
+        open_permissions_dialog("passwd")
+        b.wait_in_text("#edit-permissions-group", "admin")
+        select_access("read-write")
+        b.select_from_dropdown("#edit-permissions-group", "testuser")
+        # Switch out the file for a symlink, dialog does not update and should not dereference the symlink
+        m.execute("""
+        ln --symbolic --force /tmp/passwd /home/admin/passwd
+        chown --no-dereference admin: /home/admin/passwd""")
+        b.click(".pf-v6-c-modal-box button.pf-m-primary")
+        b.wait_not_present(".pf-v6-c-modal-box")
+        self.assertEqual(self.stat("%G", "/tmp/passwd"), "root")
+
+        # chmod lacks --no-dereference support
+        if m.image in ["rhel-8-10", "rhel-9-9", "centos-9"]:
+            self.assertNotEqual(self.stat("%A", "/tmp/passwd"), "-rw-------")
+        else:
+            self.assertEqual(self.stat("%A", "/tmp/passwd"), "-rw-------")
+
+        self.chdir('/home/admin')
+
         # As normal user you cannot change user/group permissions
         b.drop_superuser()
